src/aiida/common/hashing.py

```diff
@@ -196,27 +197,29 @@ def _(mapping, **kwargs):
 
 
 @_make_hash.register(numbers.Real)
-def _(val, **kwargs):
+def _(val: numbers.Real, **kwargs) -> list[bytes]:
     """Before hashing a float, convert to a string (via rounding) and with a fixed number of digits after the comma.
     Note that the `_single_digest` requires a bytes object so we need to encode the utf-8 string first
     """
     return [_single_digest('float', float_to_text(val, sig=AIIDA_FLOAT_PRECISION).encode('utf-8'))]
 
 
 @_make_hash.register(Decimal)
-def _(val, **kwargs):
+def _(val: Decimal, **kwargs) -> list[bytes]:
     """While a decimal can be converted exactly to a string which captures all characteristics of the underlying
     implementation, we also need compatibility with "equal" representations as int or float. Hence we are checking
     for the exponent (which is negative if there is a fractional component, 0 otherwise) and get the same hash
     as for a corresponding float or int.
     """
-    if val.as_tuple().exponent < 0:
+    # TODO: This is a possible problem!
+    # Unsupported operand types for < ("Literal['n']" and "int")
+    if val.as_tuple().exponent < 0:  # type: ignore[operator]
```

Comment:
Okay. I guess it would still be safer to guard against it here, perhaps

```suggestion
    exponent = val.as_tuple().exponent
    if isinstance(str, exponent):
        raise ValueError(f'{val} is not hashable')
    if exponent < 0:
```

Reply:
yes, I think that would be better. However, I don't think that raising is correct here. Clearly `Decimal('NaN')` is a valid value in Python, so we should be able to generate a hash for it. We can simply generate the str single_digest of the `n` exponent, right? That would result in a valid hash

Reply:
Indeed, I thought about it as well. My thinking was that if aiida does not like NaN in general it's safer to raise to prevent bugs. But if this is meant to be a general hashing function then you're correct. WDYT?

Reply:
I think that is perfectly fine. The hashing is indeed common enough that we do not have to impose limitations on acceptable values here. The limitation is in the ORM, and the check is done there, so it is fine to just accept it in the hashing implementation.

Reply:
I've added the handling of NaN et al and added a few tests. Ready for review @sphuber